### PR TITLE
Blockchain Storage Path

### PR DESCRIPTION
**Pull request by chenzhitong**
close https://github.com/neo-project/neo-express/issues/283

Can we store the blockchain file in the root directory of the program, like neo-cli, instead of store in AppData folder.

### CODE REVIEW

**Comment by chenzhitong** [comment]
@cschuchardt88 @lock9 @RichardBelsum @ixje 

Can we store the blockchain file in the root directory of the program, like neo-cli, instead of store in AppData folder.

**Comment by cschuchardt88** [comment]
Ya i can't see why not. It's annoying to have to goto `AppData` folder.

**Comment by ixje** [comment]
That sounds good.

**Comment by chenzhitong** [comment]
> Edit: But you should be using checkpoints to rollback changes and such

I don't get it. What should I do?

**Comment by cschuchardt88** [comment]
image

But we should change the directory to base directory.

**Comment by ixje** [comment]
I'm also not following your checkpoint comments. Perhaps you can describe what we're supposed to see in that screenshot and why that is (or is not) a problem?

As for where to store the data; either in the same folder as the `neo-express` configuration file, or make it configurable through the `neo-express` configuration file. And to make it backwards compatible you can say the default is "in the same folder as the config file" if not specified/present in the file.

**Comment by Liaojinghui** [comment]
Merge?

**Comment by Ashuaidehao** [comment]
May need more test to check if the `checkpoint` command can run normally after this change.
For example, I can copy the "default.neo-express" file to any project folders, then run `neoxp` commands, it can find the same folder where the chaindata stored, I'm not sure if it still work in this PR.

**Comment by chenzhitong** [comment]
> May need more test to check if the `checkpoint` command can run normally after this change. For example, I can copy the "default.neo-express" file to any project folders, then run `neoxp` commands, it can find the same folder where the chaindata stored, I'm not sure if it still work in this PR.

image

work

**Comment by cschuchardt88** [comment]
>As for where to store the data; either in the same folder as the neo-express configuration file, or make it configurable through the neo-express configuration file. And to make it backwards compatible you can say the default is "in the same folder as the config file" if not specified/present in the file.

Can we do this?

**Comment by Ashuaidehao** [comment]
> > May need more test to check if the `checkpoint` command can run normally after this change. For example, I can copy the "default.neo-express" file to any project folders, then run `neoxp` commands, it can find the same folder where the chaindata stored, I'm not sure if it still work in this PR.
> 
> image
> 
> work

It also has an offline mode, just stop the running command window and run `neoxp` in another folders.

**Comment by Ashuaidehao** [comment]
And you should test more commands like `neoxp show tx 0xXXX`, `neoxp contract deploy ...` and so on.

**Comment by chenzhitong** [comment]
I know why it's stored in the AppData folder, because it was installed through the `dotnet tool` at that time, and now it can be run as a standalone application, so it doesn't need to be stored in the `AppData` folder anymore.

PS: `dotnet tool install Neo.Express -g` are installed in the following directories by default:

| OS| Path |
| ----------- | ----------- |
| Linux/macOS | `$HOME/.dotnet/tools` |
| Windows | `%USERPROFILE%\.dotnet\tools` |

**Comment by chenzhitong** [comment]
> And you should test more commands like `neoxp show tx 0xXXX`, `neoxp contract deploy xx.nef` and so on. This PR will not just change `neoxp run` command behavior.

Test Passed

**Comment by cschuchardt88** [comment]
>I'm also not following your checkpoint comments. Perhaps you can describe what we're supposed to see in that screenshot and why that is (or is not) a problem?

I was talking about this _See Below_

>May need more test to check if the checkpoint command can run normally after this change.
For example, I can copy the "default.neo-express" file to any project folders, then run neoxp commands, it can find the same folder where the chaindata stored, I'm not sure if it still work in this PR.

**Comment by cschuchardt88** [comment]
You should change the path to match #324 Should be alot easier to understand for the user.

**Comment by Liaojinghui** [comment]
> You should change the path to match #324 Should be alot easier to understand for the user.

@chenzhitong Any update?

File: src/neoxp/Extensions/FileSystemExtensions.cs [hunk 1]
```diff
@@ -65,10 +65,7 @@ public static string GetNodePath(this IFileSystem fileSystem, ExpressConsensusNo
 
             var account = node.Wallet.Accounts.Single(a => a.IsDefault);
 
-            var rootPath = fileSystem.Path.Combine(
-                Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData, Environment.SpecialFolderOption.DoNotVerify),
-                "Neo-Express",
-                "blockchain-nodes");
+            var rootPath = "blockchain-nodes";
```
Comment (cschuchardt88) [inline src/neoxp/Extensions/FileSystemExtensions.cs:68]:
```suggestion
            var homeDir = Environment.GetFolderPath(Environment.SpecialFolder.UserProfile, Environment.SpecialFolderOption.DoNotVerify);
            var expressBlockchainDir = fileSystem.Path.Combine(homeDir, ".neo-express", "blockchain-nodes", account.ScriptHash);

            var appData = Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData, Environment.SpecialFolderOption.DoNotVerify);
            var expressOldBlockchainDir = fileSystem.Path.Combine(appData, "Neo-Express", "blockchain-nodes", account.ScriptHash);

            if (fileSystem.Directory.Exists(expressOldBlockchainDir))
                return expressOldBlockchainDir;

            return expressBlockchainDir;
```

Updated to work with #324 and be backwards compatible

File: src/neoxp/Extensions/FileSystemExtensions.cs [hunk 2]
```diff
@@ -65,10 +65,16 @@ public static string GetNodePath(this IFileSystem fileSystem, ExpressConsensusNo
...
+            var expressBlockchainDir = fileSystem.Path.Combine(homeDir, ".neo-express", "blockchain-nodes", account.ScriptHash);
+
+            var appData = Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData, Environment.SpecialFolderOption.DoNotVerify);
+            var expressOldBlockchainDir = fileSystem.Path.Combine(appData, "Neo-Express", "blockchain-nodes", account.ScriptHash);
+
+            if (fileSystem.Directory.Exists(expressOldBlockchainDir))
+                return expressOldBlockchainDir;
+
+            return expressBlockchainDir;
             return fileSystem.Path.Combine(rootPath, account.ScriptHash);
```
Comment (cschuchardt88) [inline src/neoxp/Extensions/FileSystemExtensions.cs:78]: remove this line

Comment (cschuchardt88) [inline src/neoxp/Extensions/FileSystemExtensions.cs:78]:
```suggestion

```

**Review by cschuchardt88 — APPROVED** [review]

**Review by Liaojinghui — APPROVED** [review]

**Pull request merged**